File: Sources/DomoProtocol/Identity.swift
import Foundation
import CryptoKit

/// Ed25519 keypair with a stable fingerprint identity.
/// v1 stores keys as files under DOMO_HOME; the app milestone moves the device
/// key into Keychain/Secure Enclave (see DESIGN.md §8).
public struct KeyPair {
    public let privateKey: Curve25519.Signing.PrivateKey

    public init() {
        privateKey = Curve25519.Signing.PrivateKey()
    }

    public init(rawRepresentation: Data) throws {
        privateKey = try Curve25519.Signing.PrivateKey(rawRepresentation: rawRepresentation)
    }

    public var publicKeyBase64: String {
        privateKey.publicKey.rawRepresentation.base64EncodedString()
    }

    public var privateKeyBase64: String {
        privateKey.rawRepresentation.base64EncodedString()
    }

    /// Stable short identity derived from the public key.
    public var fingerprint: String {
        Self.fingerprint(ofPublicKeyBase64: publicKeyBase64)
    }

    public static func fingerprint(ofPublicKeyBase64 publicKey: String) -> String {
        let digest = SHA256.hash(data: Data(publicKey.utf8))
        return digest.prefix(8).map { String(format: "%02x", $0) }.joined()
    }

    public func sign(_ data: Data) throws -> Data {
        try privateKey.signature(for: data)
    }

    public static func verify(signature: Data, data: Data, publicKeyBase64: String) -> Bool {
        guard let keyData = Data(base64Encoded: publicKeyBase64),
              let key = try? Curve25519.Signing.PublicKey(rawRepresentation: keyData) else {
            return false
        }
        return key.isValidSignature(signature, for: data)
    }
}

public enum Hashing {
    public static func sha256Hex(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }
}

File: packages/protocol/src/identity.ts
/**
 * Ed25519 keypair with a stable fingerprint identity — twin of
 * DomoProtocol/Identity.swift. Node's crypto exposes Ed25519 through
 * DER-wrapped keys; CryptoKit uses raw 32-byte representations. The fixed
 * DER prefixes below convert between them (constants of the algorithm).
 * Ed25519 signatures are deterministic (RFC 8032), so both implementations
 * produce byte-identical signatures — asserted by fixtures/identity.json.
 */
import crypto from "node:crypto";

const PKCS8_ED25519_PREFIX = Buffer.from("302e020100300506032b657004220420", "hex");
const SPKI_ED25519_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

function privateKeyFromSeed(seed: Buffer): crypto.KeyObject {
  return crypto.createPrivateKey({
    key: Buffer.concat([PKCS8_ED25519_PREFIX, seed]),
    format: "der",
    type: "pkcs8",
  });
}

function publicKeyFromRaw(raw: Buffer): crypto.KeyObject {
  return crypto.createPublicKey({
    key: Buffer.concat([SPKI_ED25519_PREFIX, raw]),
    format: "der",
    type: "spki",
  });
}

export class KeyPair {
  private readonly privateKey: crypto.KeyObject;
  private readonly seed: Buffer;
  readonly publicKeyRaw: Buffer;

  constructor(seed?: Buffer) {
    this.seed = seed ?? crypto.randomBytes(32);
    if (this.seed.length !== 32) throw new Error("Ed25519 seed must be 32 bytes");
    this.privateKey = privateKeyFromSeed(this.seed);
    const spki = crypto
      .createPublicKey(this.privateKey)
      .export({ format: "der", type: "spki" }) as Buffer;
    this.publicKeyRaw = spki.subarray(SPKI_ED25519_PREFIX.length);
  }

  static fromRawRepresentation(raw: Buffer): KeyPair {
    return new KeyPair(raw);
  }

  get publicKeyBase64(): string {
    return this.publicKeyRaw.toString("base64");
  }

  get privateKeyBase64(): string {
    return this.seed.toString("base64");
  }

  /** Stable short identity derived from the public key. */
  get fingerprint(): string {
    return KeyPair.fingerprintOfPublicKeyBase64(this.publicKeyBase64);
  }

  static fingerprintOfPublicKeyBase64(publicKeyBase64: string): string {
    return crypto
      .createHash("sha256")
      .update(Buffer.from(publicKeyBase64, "utf8"))
      .digest()
      .subarray(0, 8)
      .toString("hex");
  }

  sign(data: Buffer): Buffer {
    return crypto.sign(null, data, this.privateKey);
  }

  static verify(signature: Buffer, data: Buffer, publicKeyBase64: string): boolean {
    let key: crypto.KeyObject;
    try {
      const raw = Buffer.from(publicKeyBase64, "base64");
      if (raw.length !== 32) return false;
      key = publicKeyFromRaw(raw);
    } catch {
      return false;
    }
    try {
      return crypto.verify(null, data, key, signature);
    } catch {
      return false;
    }
  }
}

export const Hashing = {
  sha256Hex(data: Buffer): string {
    return crypto.createHash("sha256").update(data).digest("hex");
  },
};
